### PR TITLE
Handle Multiple Stacks

### LINKED ISSUE #289: Multiple stacks
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

```
cdktf --version
0.0.13-pre.2f0b966cb25b6dc5066ee1551fdb32ff47a9c0fd
```

### Affected Resource(s)

Multiple stacks.

### Expected Behavior

I'm trying to see if I can use multiple stacks in multiple files like in CDK.
This is my current program

file1.ts 

```typescript
import {IamRole} from "./.gen/providers/aws";
import {TerraformStack} from 'cdktf'
import {Construct} from "constructs";


export class SecStack extends TerraformStack {
    constructor(scope: Construct, id: string) {
        super(scope, id);
        new IamRole(this, "cdk-tf-module-role", {
            name: "cdk-tf-module-role",
            assumeRolePolicy: JSON.stringify({
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Action": "sts:AssumeRole",
                        "Principal": {
                            "Service": "lambda.amazonaws.com"
                        },
                        "Effect": "Allow",
                    }
                ]
            })
        })
    }
}
```

file2.ts
```typescript
import {Construct} from 'constructs';
import {App, TerraformOutput, TerraformStack} from 'cdktf';
import {AwsProvider, LambdaFunction, IamRole} from "./.gen/providers/aws";
import {SecStack} from "./role";

class MainStack extends TerraformStack {
    constructor(scope: Construct, id: string) {
        super(scope, id);

        new AwsProvider(this, "aws", {
            region: "eu-west-1",
        });

        const lambdaRole = new IamRole(this, "cdk-tf-lambda-role", {
            name: "cdk-tf-lambda-role",
            assumeRolePolicy: JSON.stringify({
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Action": "sts:AssumeRole",
                        "Principal": {
                            "Service": "lambda.amazonaws.com"
                        },
                        "Effect": "Allow",
                    }
                ]
            })
        })

        const lambda = new LambdaFunction(this, "cdk-tf-lambda", {
            filename: '../lambda/lambda.zip',
            functionName: 'cdk-tf',
            handler: 'index.handler',
            role: lambdaRole.arn,
            runtime: 'nodejs12.x'
        });

        new TerraformOutput(this, 'cdk-tf-lambda-arn', {
            value: lambda.arn
        });
    }
}

const app = new App();
new MainStack(app, 'dec-imp-main-stack');
new SecStack(app, 'dec-imp-sec-stack')
app.synth();
```

When `cdktf deploy` it appears that only the second stack is `synth`

cdk.tf.json
```
{
  "//": {
    "metadata": {
      "version": "0.0.13-pre.2f0b966cb25b6dc5066ee1551fdb32ff47a9c0fd",
      "stackName": "dec-imp-sec-stack"
    }
  },
  "resource": {
    "aws_iam_role": {
      "decimpsecstack_cdktfmodulerole_72EFC78B": {
        "assume_role_policy": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":\"sts:AssumeRole\",\"Principal\":{\"Service\":\"lambda.amazonaws.com\"},\"Effect\":\"Allow\"}]}",
        "name": "cdk-tf-module-role",
        "//": {
          "metadata": {
            "path": "dec-imp-sec-stack/cdk-tf-module-role",
            "uniqueId": "decimpsecstack_cdktfmodulerole_72EFC78B",
            "stackTrace": [
              "new TerraformElement (/Users/thomas.schersach/code/dec-imp/node_modules/cdktf/lib/terraform-element.js:11:28)",
              "new TerraformResource (/Users/thomas.schersach/code/dec-imp/node_modules/cdktf/lib/terraform-resource.js:9:9)",
              "new IamRole (/Users/thomas.schersach/code/dec-imp/.gen/providers/aws/iam-role.js:13:9)",
              "new SecStack (/Users/thomas.schersach/code/dec-imp/role.js:9:9)",
              "Object.<anonymous> (/Users/thomas.schersach/code/dec-imp/main.js:48:1)",
              "Module._compile (internal/modules/cjs/loader.js:1137:30)",
              "Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)",
              "Module.load (internal/modules/cjs/loader.js:985:32)",
              "Function.Module._load (internal/modules/cjs/loader.js:878:14)",
              "Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)",
              "internal/main/run_main_module.js:17:47"
            ]
          }
        }
      }
    }
  }
}
```

### Actual Behavior

The deploy is running forever (plan stage), it doesn't seem to work.

### Steps to Reproduce

See program

### PR DESCRIPTION
**Pull request by skorfmann**
This will change the commands `synth`, `diff`, `deploy` and `destroy` to handle more than one stack per application.

## Current Behaviour prior to this Change 

When using `cdktf` the application code is synthesized into the configured output folder (`cdktf.out` by default) as `cdk.tf.json` file. The output folder serves as working directory for the Terraform CLI for the commands which interacting with it (`diff`, `deploy`, `destroy`). Since we do have a static output folder, users might copy assets such as zip files as part of the stack synth manually in the output folder.

## Changed behaviour 

We'll have a `stacks` folder within the app output directory (`cdktf.out`), where each stack will synthesize itself in a sub directory (e.g. `cdktf.out/stacks/stack-a`). This will then serve as the working directory for all Terraform CLI commands. There won't be an automatic cleanup of the stacks directory.

## Planned Changes

- [x] Synth all stacks at once 
- [x] Synth a single stack only (useful when dealing with stacks which process large assets)
- [x] Adapt `deploy`, `diff`, `destroy` to take a positional stack argument. If more than one stack exists, that's a mandatory argument 
- [x] Add a list stacks command 
- [x] Make sure the Terraform Cloud remote exec workflow doesn't break. This won't manage the Terraform workspaces on behalf of the user and therefore expects the workspaces to be present.
- [x] ~~Find a good way to access the stack's working directory on a construct level (e.g. to copy / generate assets)~~ (extracted in #648)
- [x] Add / change documentation for stacks. Point out, that we don't support automatic cross-stack references at this point
- [x] Make sure local workflows handle the state files correctly 
- [x] Add check for existing `terraform.tfstate` file in project root and prompt the user to rename it 

Fixes #289 #35

### CODE REVIEW

**Comment by skorfmann** [comment]
> * Find a good way to access the stack's working directory on a construct level (e.g. to copy / generate assets)

This is possible in the current state, but not really convenient. Will create a follow up issue to make this more idiomatic.

File: test/typescript/multiple-stacks/test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+      Diff: 1 to create, 0 to update, 0 to delete.
+      "
+    `);
+
+
+    let error;
+    try { driver.diff() } catch(e) { error = e.message }
+    expect(error).toMatch('Found more than one stack, please specify a target stack first, second');
```
Comment (ansgarm) [inline test/typescript/multiple-stacks/test.ts:52]: Could probably simplified to something like `expect(() => driver.diff()).toThrowError('Found more than …')` [docs](https://jestjs.io/docs/expect#tothrowerror)

File: packages/cdktf-cli/bin/cmds/helper/synth-stack.ts [hunk 2]
```diff
@@ -57,28 +69,28 @@ Command output on stderr:
...
       console.error('ERROR: No Terraform code synthesized.');
     }
 
-    if (stacks.length > 1) {
-      console.error('ERROR: Found more than one stack. Multiple stacks are not supported at the moment and might lead to unpredictable behaviour.');
+    const stackNames = stacks.map(s => s.name)
+    const orphanedDirectories = existingDirectories.filter(e => !stackNames.includes(path.basename(e)))
+
+    for (const orphanedDirectory of orphanedDirectories) {
+      fs.rmdirSync(orphanedDirectory, { recursive: true })
```
Comment (ansgarm) [inline packages/cdktf-cli/bin/cmds/helper/synth-stack.ts:93]: Could this somehow lead to unexpected side effects like loosing local Terraform state after renaming a stack?

**Comment by ansgarm** [comment]
Maybe we should also update the `help` file in our templates to contain the (sometimes optional) `stack` argument for some commands?

File: packages/cdktf-cli/bin/cmds/helper/synth-stack.ts (hunk 2 above)
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/helper/synth-stack.ts:93]: Just checked. The local Terraform state is writting to the root directory and not touched 👍

**Comment by ansgarm** [comment]
Just tested it locally, works great 🚀

File: packages/cdktf-cli/bin/cmds/list.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import { displayVersionMessage } from './version-check'
+
+const config = readConfigSync();
+
+class Command implements yargs.CommandModule {
+  public readonly command = 'list [OPTIONS]';
+  public readonly describe = 'List stacks in app.';
+
+  public readonly builder = (args: yargs.Argv) => args
+    .option('app', { default: config.app, desc: 'Command to use in order to execute cdktf app', alias: 'a' })
```
[the remark this answers is not in the document]
Comment (skorfmann) [reply packages/cdktf-cli/bin/cmds/list.ts:15]: Do you have a suggestion for a better wording?

Haven't thought much about this so far, it was derived from [cdk8s](https://github.com/cdk8s-team/cdk8s/blob/bf9d0d5351cdb206a6347ae99967e24ab4e46f47/packages/cdk8s-cli/src/cli/cmds/synth.ts#L15) which probably took it from the AWS CDK :)

File: packages/cdktf-cli/bin/cmds/terraform-check.ts [hunk 4]
```diff
@@ -1,12 +1,27 @@
...
+import * as path from 'path';
 
 const MIN_SUPPORTED_VERSION = '0.13.0'
 const VERSION_REGEXP = /Terraform v\d+.\d+.\d+/
 
 export const terraformCheck = async (): Promise<void> => {
   try {
-    const terraform = new TerraformCli('./')
+    if (existsSync(path.join(process.cwd(), 'terraform.tfstate'))) {
+      throw new Error(`Found 'terraform.tfstate' Terraform state file. Please rename it to match the stack name. Learn more https://cdk.tf/multiple-stacks`)
```
[the remark this answers is not in the document]
Comment (skorfmann) [reply packages/cdktf-cli/bin/cmds/terraform-check.ts:13]: That's a great point! 🎉

File: packages/cdktf-cli/bin/cmds/list.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/bin/cmds/list.ts:15]: Just a thought (unsure if it's better)
> Command to use in order to execute cdktf app entrypoint
Feel free to resolve however 😄

[the remark this answers is not in the document]
Comment (skorfmann) [reply packages/cdktf-cli/bin/cmds/list.ts:15]: Will keep it as is 👍

**Review by ansgarm — APPROVED** [review]

File: test/csharp/synth-app/test.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ describe("csharp full integration test synth", () => {
   });
 
   test("synth generates JSON", async () => {
-    driver.synth()
+    await driver.synth()
     expect(driver.synthesizedStack('csharp-simple')).toMatchSnapshot()
-  })
+  }, 180_000);
```
[the remark this answers is not in the document]
Comment (ansgarm) [reply test/csharp/synth-app/test.ts:19]: Oh, it seems to be there now. I must've reviewed some intermediate state or something..

File: packages/cdktf-cli/bin/cmds/destroy.ts [hunk 6]
```diff
@@ -12,7 +12,7 @@ class Command implements yargs.CommandModule {
   public readonly describe = 'Destroy the given stack';
 
   public readonly builder = (args: yargs.Argv) => args
-    .positional('stack', { desc: 'Destroy stack which matches the given id only. Required when more than stack is present in the app', type: 'string' })
+    .positional('stack', { desc: 'Destroy stack which matches the given id only. Required when more than one stack  is present in the app', type: 'string' })
```
[the remark this answers is not in the document]
Comment (skorfmann) [reply packages/cdktf-cli/bin/cmds/destroy.ts:15]: Great attention to detail :) Fixed it

**Pull request merged**